internals/plan/extensions_test.go

```diff
@@ -0,0 +1,687 @@
+		extensions: map[string]plan.LayerSectionExtension{
+			"x-field": &xExtension{},
+			"y-field": &yExtension{},
+		},
+		layers: []*inputLayer{
+			{
+				order: 1,
+				label: "layer-x",
+				yaml: `
+					summary: x
+					description: desc-x
+					x-field:
+						x1:
+							override: replace
+							a: a
+							b: b
+							y-field:
+							  - y1
+				`,
+			},
+			{
+				order: 2,
+				label: "layer-y",
+				yaml: `
+					summary: y
+					description: desc-y
+					y-field:
+						y1:
+							override: replace
+							a: a
+							b: b
+				`,
+			},
+		},
+		result: &planResult{
+			x: &xSection{
+				Entries: map[string]*X{
+					"x1": &X{
+						Name:     "x1",
+						Override: plan.ReplaceOverride,
+						A:        "a",
+						B:        "b",
+						Y: []string{
+							"y1",
+						},
+					},
+				},
+			},
+			y: &ySection{
+				Entries: map[string]*Y{
+					"y1": &Y{
+						Name:     "y1",
+						Override: plan.ReplaceOverride,
+						A:        "a",
+						B:        "b",
+					},
+				},
+			},
+		},
+		resultYaml: string(reindent(`
+			x-field:
+				x1:
+					override: replace
+					a: a
+					b: b
+					y-field:
+						- y1
+			y-field:
+				y1:
+					override: replace
+					a: a
+					b: b`)),
+	},
+}
+
+func (s *S) TestPlanExtensions(c *C) {
+nexttest:
+	for testIndex, testData := range extensionTests {
+		c.Logf("TestPlanExtensions :: %s (data index %v)", testData.summary, testIndex)
+
+		// Write layers to test directory.
+		layersDir := filepath.Join(c.MkDir(), "layers")
+		s.writeLayerFiles(c, layersDir, testData.layers)
+		var p *plan.Plan
+
+		// Register test extensions.
+		for field, extension := range testData.extensions {
+			err := func() (err error) {
+				defer func() {
+					if r := recover(); r != nil {
+						err = fmt.Errorf("%v", r)
+					}
+				}()
+				plan.RegisterExtension(field, extension)
+				return nil
+			}()
+			if err != nil {
+				c.Assert(err, ErrorMatches, testData.error)
+				continue nexttest
+			}
+		}
+
+		// Load the plan layer from disk (parse, combine and validate).
+		p, err := plan.ReadDir(layersDir)
+		if testData.error != "" || err != nil {
+			// Expected error.
+			c.Assert(err, ErrorMatches, testData.error)
+		} else {
+			if _, ok := testData.extensions[xField]; ok {
+				// Verify "x-field" data.
+				var x *xSection
+				x = p.Section(xField).(*xSection)
+				c.Assert(err, IsNil)
+				c.Assert(x.Entries, DeepEquals, testData.result.x.Entries)
+			}
+
+			if _, ok := testData.extensions[yField]; ok {
+				// Verify "y-field" data.
+				var y *ySection
+				y = p.Section(yField).(*ySection)
+				c.Assert(err, IsNil)
+				c.Assert(y.Entries, DeepEquals, testData.result.y.Entries)
+			}
+
+			// Verify combined plan YAML.
+			planYAML, err := yaml.Marshal(p)
+			c.Assert(err, IsNil)
+			c.Assert(string(planYAML), Equals, testData.resultYaml)
+		}
+
+		// Unregister test extensions.
+		for field, _ := range testData.extensions {
+			plan.UnregisterExtension(field)
```

Comment:
This cannot be here, I believe, as any Assert above will leave the runtime in a bad state for further tests. Such cases need to be on a defer, or as part of the suite tear down functions.

Reply:
I fixed this so that extension removal now happens irrespective of the test iteration result. Resolving.